File: 20171030-kotlin-game-of-life/src/test/kotlin/com/katas/CellSpec.kt
package com.katas

import org.assertj.core.api.Assertions.assertThat
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it

object CellSpec : Spek({

    describe("A live Cell") {
        it("dies when has less than two alive neighbours") {
            val cell = Cell(true, listOf(Cell(true)))

            assertThat(cell.nextGeneration().alive).isFalse()
        }

        it("survives when has two alive neighbours") {
            val cell = Cell(true, listOf(Cell(true), Cell(true)))

            assertThat(cell.nextGeneration().alive).isTrue()
        }

        it("survives when has three alive neighbours") {
            val cell = Cell(true, listOf(Cell(true), Cell(true), Cell(true)))

            assertThat(cell.nextGeneration().alive).isTrue()
        }

        it("dies when has more than three alive neighbours") {
            val cell = Cell(true, listOf(Cell(true), Cell(true), Cell(true), Cell(true)))

            assertThat(cell.nextGeneration().alive).isFalse()
        }

    }

    describe("A dead Cell") {

        it("stays dead next generation") {
            val cell = Cell(false, listOf(Cell(true), Cell(true)))

            assertThat(cell.nextGeneration().alive).isFalse()
        }

        it("resurrects when has three alive neighbours") {
            val cell = Cell(false, listOf(Cell(true), Cell(true), Cell(true)))

            assertThat(cell.nextGeneration().alive).isTrue()
        }

    }

})

File: 20181111-ts-gol/spec/cell.spec.ts
import { Cell } from '../lib/cell';

describe('Cell', () => {

    let cell: Cell;

    beforeEach(() => {
        cell = new Cell(0, 0, true);
    });

    it('should die when has less than 2 neighbours', () => {
        const neighbours = [new Cell(0, 1, true)];

        cell.tick(neighbours);

        expect(cell.isAlive).toBe(false);
    });

    it('should continue living if is alive and has two alive neighbours', () => {
        const neighbours = [new Cell(0, 1, true), new Cell(1, 0, true)];

        cell.tick(neighbours);

        expect(cell.isAlive).toBe(true);
    });

    it('should continue living if is alive and has three alive neighbours', () => {
        const neighbours = [new Cell(0, 1, true),
                            new Cell(1, 0, true),
                            new Cell(1, 1, true)];

        cell.tick(neighbours);

        expect(cell.isAlive).toBe(true);
    });

    it('should die by overpopulation if is alive and has more than 3 alive neighbours', () => {
        const cell = new Cell(1, 1, true);
        const neighbours = [new Cell(0, 0, true),
                            new Cell(0, 1, true),
                            new Cell(0, 2, true),
                            new Cell(1, 0, true)];

        cell.tick(neighbours);

        expect(cell.isAlive).toBe(false);
    });

    it('should become live if has exactly three alive neighbours', () => {
        const cell = new Cell(0, 0, false);
        const neighbours = [new Cell(0, 1, true),
                            new Cell(1, 0, true),
                            new Cell(1, 1, true)];

        cell.tick(neighbours);

        expect(cell.isAlive).toBe(true);
    });
});